### ISSUE TITLE
Type narrowing doesn't work inside `if type(a) is b.__class__:`

### ISSUE THREAD

**Issue by gdiscry** [microsoft/pyright#3045]
**Describe the bug**
Type narrowing doesn't work inside `if type(a) is b.__class__:`, while it works inside `if type(a) is B:` or `if isinstance(a, b.__class__):`.

**To Reproduce**
When type-checking the code below, the following errors are reported:
```
foo.py
  foo.py:8:25 - information: Type of "other" is "object"
  foo.py:9:39 - error: Cannot access member "attr" for type "object"
    Member "attr" is unknown (reportGeneralTypeIssues)
  foo.py:9:33 - error: Type of "attr" is unknown (reportUnknownMemberType)
  foo.py:9:20 - error: Return type is unknown (reportUnknownVariableType)
3 errors, 0 warnings, 1 information 
```

**Expected behavior**
There shouldn't be any type error with that code and `reveal_type(other)` should display `Self@A` (probably?) instead of `object`.

**Screenshots or Code**
```python
# pyright: strict

class A:
    attr: object

    def __eq__(self, other: object) -> bool:
        if type(other) is self.__class__:
            reveal_type(other)
            return self.attr == other.attr
        return NotImplemented
```

**VS Code extension or command-line**
I am running pyright as a command-line tool:
```console
$ pyright --version             
pyright 1.1.220
```

**Additional context**
I am implementing some base classes that shouldn't be comparable with their subclasses. Any of the special methods used for comparison should return `NotImplemented` in that case. This is similar to what attrs does:
```python
    def __eq__(self, other):
        if other.__class__ is not self.__class__:
            return NotImplemented
        ...
```
but with a form that should be supported by several type checkers for type narrowing. Mypy actually narrows the type like expected.

**Comment by erictraut** [comment]
Thanks for reporting the issue. The type narrowing logic for the `type(x) is y` pattern wasn't handling the case where `y` was a TypeVar or Self type. I've enhanced the logic to handle this case. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.222, which I just published. It will also be included in the next release of pylance.

**Comment by gdiscry** [comment]
Thanks for the fix, I can confirm it solved my issue.
